File: py_scripts/code/scripts/forest.py
import copy
from typing import List
import numpy as np
from numpy.core import ndarray
from scripts.hg import Histogram
import matplotlib.pyplot as plt
from math import log

to_print = False


class Forest:
    """This creates a forest of trees, given the following list of parameters:
    1) n_trees: the number of trees
    2) max_depth: the depth of the trees
    3) max_samples: number of samples per tree
    4) max_buckets: maximum number of buckets used by the histogram
    5) epsilon: accuracy of the histogram"""
    dim = ...  # type: int
    size = ...  # type: int
    points = ...  # type: ndarray
    start = ...  # type: ndarray
    end = ...  # type: ndarray

    
    def __init__(self, **kwargs):
        self.n_trees = kwargs['n_trees']
        self.max_depth = kwargs['max_depth']
        self.max_samples = kwargs['max_samples']
        self.max_buckets = kwargs['max_buckets']
        self.epsilon = kwargs['epsilon']
        self.sample_axis = kwargs['sample_axis']
        self.threshold = kwargs['threshold']
        self.tree = []
        self.n_leaves = np.zeros(self.n_trees)

    
    def fit(self, pts):
        self.points = pts
        self.dim, self.size = np.shape(self.points)
        print("Dim ", self.dim, "Size ", self.size)
        if int(self.sample_axis*self.dim) == 0:
            print("sample_axis is too low")
            return
        self.start = np.zeros(self.dim)
        self.end = np.zeros(self.dim)
        print("Init start dim, Init end dim")
        print(self.start, self.end)
        
        for axis in range(self.dim):
            val = np.unique(np.array(self.points[axis]))
            if len(val) <= 1:
                print("No entropy in dimension :", axis)
                return
            self.start[axis] = (3 * val[0] - val[1]) / 2
            self.end[axis] = (3 * val[-1] - val[-2]) / 2

        print("Final start dim, Final end dim")
        print(self.start, self.end)
        
        k_args = {'depth': 0, 'forest': self}
        max_sample_size = np.min((self.size, self.max_depth*200))
        sample = np.random.choice(self.size, max_sample_size, replace=False)
        for i in range(self.n_trees):
            print("Building tree #", i)
            k_args['indices'] = np.random.choice(self.size, self.max_samples, replace=False)
            root_node = Node(**k_args)
            root_node.compute_density(sample)
            self.tree.append(root_node)
            self.n_leaves[i] = root_node.compute_leaf_num()

    def predict(self, pts, err=0.1, pct=50):
        _, n_pts = np.shape(pts)
        print("n_pts ", n_pts)
        scores = np.zeros((self.n_trees, n_pts))
        indices = [i for i in range(n_pts)]
        for i in range(self.n_trees):
            self.tree[i].compute_split(pts, indices, scores[i])
        n_err = int(err * n_pts)
        min_score = np.percentile(scores, pct, axis=0)
        top_indices = np.argsort(min_score)[:n_err]
        anom_pts = {}
        anom_scores = {}
        anom_pct = {}
        for i in range(n_err):
            anom_pts[top_indices[i]] = pts[:, top_indices[i]]
            anom_scores[top_indices[i]] = scores[:, top_indices[i]]
            anom_pct[top_indices[i]] = min_score[top_indices[i]]
        return top_indices, anom_pts, anom_scores, anom_pct, min_score

class PointSet:
    
    def __init__(self, node, indices):
        self.node = node
        self.indices = indices
        self.val = []
        self.count = []
        self.gap = []
        for axis in range(self.node.forest.dim):
            val, count = np.unique(np.array(self.node.forest.points[axis, self.indices]), return_counts=True)
            self.val.append(val)
            self.count.append(count)
            if len(val) <= 1:
                gap = [0]
            else:
                val_arr = np.array(val)
                gap_arr = np.zeros(len(val))
                gap_arr[0] = (val_arr[1] - val_arr[0]) / 2
                gap_arr[-1] = (val_arr[-1] - val_arr[-2]) / 2
                gap_arr[1:-1] = (val_arr[2:] - val_arr[:-2]) / 2
                gap = gap_arr.tolist()
            self.gap.append(gap)


class Cube:
    
    def __init__(self, node, start, end):
        assert isinstance(node, Node)
        self.node = node
        self.child = []
        self.start = start
        self.end = end
        self.dim = len(start)
        self.split_axis = -1
        self.split_vals = []
        self.vol = 0
        for i in range(self.dim):
            self.vol += log(self.end[i] - self.start[i])
    
    def filter_indices(self, indices):
        in_lb = self.node.forest.points[:, indices] >= self.start.reshape(self.dim, 1)
        in_ub = self.node.forest.points[:, indices] < self.end.reshape(self.dim, 1)
        return [indices[i] for i in range(len(indices)) if in_lb[:, i].all() and in_ub[:, i].all()]
    
    def split_indices(self, pts, indices): # original
        n_child = len(self.child)
        if n_child == 0:
            return indices
        n_arr = len(indices)
        if n_arr == 0:
            return [[] for _ in range(n_child)]
        s_arr = pts[self.split_axis]
        s_start = self.start[self.split_axis]
        s_end = self.end[self.split_axis]
        index_split = [[] for _ in range(n_child)]
        index_split[0] = [ind for ind in indices if ((s_arr[ind] >= s_start) and (s_arr[ind] < self.split_vals[0]))]
        index_split[-1] = [ind for ind in indices if ((s_arr[ind] >= self.split_vals[-1]) and (s_arr[ind] < s_end))]
        for k in range(1, n_child - 1):
            index_split[k] = [ind for ind in indices if (s_arr[ind] >= self.split_vals[k - 1]) and
                              (s_arr[ind] < self.split_vals[k])]
        return index_split
    

class Node:

    
    def __init__(self, depth, forest, **kwargs):
        self.depth = depth
        self.forest = forest
        if self.depth == 0:
            self.id_string = [0]
            self.cube = Cube(self, self.forest.start, self.forest.end)
            self.point_set = PointSet(self, kwargs['indices'])
        else:
            self.id_string = kwargs['id']
            self.cube = Cube(self, kwargs['start'], kwargs['end'])
            self.point_set = PointSet(self, self.cube.filter_indices(kwargs['indices']))
        self.density = -1
        self.child = []
        if (self.depth < self.forest.max_depth) and (len(self.point_set.indices) > 1):
            self.find_split()

    
    def find_split(self):
        print("self.cube.dim, ", self.cube.dim)
        imp_axis = [axis for axis in range(self.cube.dim) if len(self.point_set.val[axis]) > 1]
        if not imp_axis:
            return
        max_axes = min(len(imp_axis), int(self.forest.sample_axis * self.cube.dim))
        s_axes = np.random.choice(imp_axis, max_axes, replace=False)
        print("s_axes, ", s_axes)
        
        buckets = {}
        var_red = {}
        for axis in s_axes:
            hist = Histogram(self.point_set.gap[axis] / self.point_set.count[axis], self.point_set.count[axis],
                             self.forest.max_buckets, self.forest.epsilon)
            _, var_red[axis], buckets[axis] = hist.best_split()
            print("axis, var_red[axis], buckets[axis]: ", axis, ", ", var_red[axis], ", ", buckets[axis])
            
        if np.max(list(var_red.values())) <= self.forest.threshold:
            return

        split_axis = np.random.choice(s_axes, p=list(var_red.values()) / np.sum(list(var_red.values())))
        self.cube.split_axis = split_axis
        self.cube.split_vals = [(self.point_set.val[split_axis][i - 1] + self.point_set.val[split_axis][i]) / 2 for i in
                                buckets[split_axis]]
        for i in range(len(self.cube.split_vals) + 1):
            new_start = np.array(self.cube.start)
            new_end = np.array(self.cube.end)
            if 0 < i < len(self.cube.split_vals):
                new_start[split_axis] = self.cube.split_vals[i - 1]
                new_end[split_axis] = self.cube.split_vals[i]
            elif i == 0:
                new_end[split_axis] = self.cube.split_vals[0]
            else: 
                new_start[split_axis] = self.cube.split_vals[-1]
            new_id = copy.deepcopy(self.id_string)
            new_id.append(i)
            kwargs = {'start': new_start, 'end': new_end}
            kwargs.update({'indices': self.point_set.indices, 'id': new_id})
            child_node = Node(self.depth + 1, self.forest, **kwargs)
            self.child.append(child_node)
            self.cube.child.append(child_node.cube)

    
    def compute_density(self, indices):
        num = len(indices)
        if num == 0:
            self.density = 0
            self.child = []
            self.cube.child = []
            self.cube.split_axis = -1
            return
        self.density = log(num) - self.cube.vol
        if self.child:
            index_split = self.cube.split_indices(self.forest.points, indices)
            print("index_split (sizes) [NODE]: ", [len(n) for n in index_split])
            for i in range(len(self.child)):
                self.child[i].compute_density(index_split[i])

    
    def compute_leaf_num(self):
        if self.child:
            leaves = 0
            for i in range(len(self.child)):
                leaves += self.child[i].compute_leaf_num()
            return leaves
        else:
            return 1

    
    def compute_split(self, pts, indices, scores):
        if self.child:
            index_split = self.cube.split_indices(pts, indices)
            for i in range(len(self.child)):
                if len(index_split[i]) > 0:
                    self.child[i].compute_split(pts, index_split[i], scores)
        else:
            scores[indices] = self.density

File: cpp_scripts/new/code/Forest.cpp
#include "Forest.h"
#include "Node.h"
#include <set>

Forest::Forest(const std::unordered_map<std::string, double>& kwargs) {
    n_trees = static_cast<int>(kwargs.at("n_trees"));
    max_depth = static_cast<int>(kwargs.at("max_depth"));
    max_samples = static_cast<int>(kwargs.at("max_samples"));
    max_buckets = static_cast<int>(kwargs.at("max_buckets"));
    epsilon = kwargs.at("epsilon");
    sample_axis = kwargs.at("sample_axis");
    threshold = kwargs.at("threshold");
    tree.reserve(n_trees);
    n_leaves.resize(n_trees);
}

void Forest::fit(const std::vector<std::vector<double>>& pts) {
    points = pts;
    size = static_cast<int>(points.size());
    dim = static_cast<int>(points[0].size());
    std::cout << "Dim " << dim << ", Size " << size << std::endl;
    if (sample_axis * dim < 1) {
        std::cout << "sample_axis is too low" << std::endl;
        return;
    }
    start.resize(dim);
    end.resize(dim);
    
    std::vector<std::vector<double>> transposed_points(dim, std::vector<double>(size));
    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < dim; ++j) {
            transposed_points[j][i] = points[i][j];
        }
    }

    for (int axis = 0; axis < dim; ++axis) {
        std::vector<double> val(transposed_points[axis].begin(), transposed_points[axis].end());
        std::sort(val.begin(), val.end());
        val.erase(std::unique(val.begin(), val.end()), val.end());
        start[axis] = (3 * val[0] - val[1]) / 2;
        end[axis] = (3 * val.back() - val[val.size() - 2]) / 2;
    }
    start.resize(dim);
    end.resize(dim);
    std::vector<int> sample(std::min(size, max_depth * 200));
    std::iota(sample.begin(), sample.end(), 0);
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(sample.begin(), sample.end(), g);
    std::cout << "\n >> Building trees\n";
    for (int i = 0; i < n_trees; ++i) {
        
        std::vector<int> indices(max_samples);
        std::iota(indices.begin(), indices.end(), 0);
        std::random_device rd;
        std::mt19937 g(rd());
        std::shuffle(indices.begin(), indices.end(), g);
        std::cout << "\n >> Building tree #" << i + 1<< "/" << n_trees <<"\n";
        Node root_node(0, this, {{"indices", indices}});
        root_node.compute_density(sample);
        tree.push_back(root_node);
        n_leaves[i] = root_node.compute_leaf_num();
        std::cout << "\n >> Building tree #" << i + 1 << "/" << n_trees <<" [DONE]\n";
    }
    std::cout << "\n >> Building trees [DONE]\n";    
}

std::tuple<std::vector<int>, std::unordered_map<int, std::vector<double>>,
        std::unordered_map<int, std::vector<double>>, std::unordered_map<int, double>, std::vector<double>>
Forest::predict(const std::vector<std::vector<double>>& pts, double err, double pct) {
    
    int n_pts = static_cast<int>(pts.size());
    std::vector<std::vector<double>> scores(n_trees, std::vector<double>(n_pts));
    std::vector<int> indices;
    indices.reserve(n_pts);
    for (int i = 0; i < n_pts; i++) {
        indices.push_back(i);
    }

    for (int i = 0; i < n_trees; ++i) {
        tree[i].compute_split(pts, indices, scores[i]);
    }
    int n_err = static_cast<int>(err * n_pts);
    
    std::vector<double> min_score(n_pts);
    for (int i = 0; i < n_pts; ++i) {
        std::vector<double> column_scores(n_trees);
        bool all_zeroes = 1;
        for (int j = 0; j < n_trees; ++j) {
            column_scores[j] = scores[j][i];
            if (column_scores[j] > 0 || column_scores[j] < 0) {
                all_zeroes = 0;
            }
        }
        std::sort(column_scores.begin(), column_scores.end());
        int index = static_cast<int>(pct / 100.0 * (n_trees - 1)); // since pct = 0, this always sets index to 0 (min)
        min_score[i] = column_scores[index];
    }
    
    std::vector<int> top_indices(n_pts);
    std::iota(top_indices.begin(), top_indices.end(), 0);

    std::sort(top_indices.begin(), top_indices.end(), [&](int a, int b) {
        return min_score[a] < min_score[b];
    });

    top_indices.resize(n_err);

    std::unordered_map<int, std::vector<double>> anom_pts;
    std::unordered_map<int, std::vector<double>> anom_scores;
    std::unordered_map<int, double> anom_pct;

    for (int i = 0; i < n_err; ++i) {
        int index = top_indices[i];
        anom_pts[index] = std::vector<double>(n_trees);
        anom_scores[index] = std::vector<double>(n_trees);
        for (int j = 0; j < n_trees; ++j) {
            anom_pts[index][j] = pts[j][index];
            anom_scores[index][j] = scores[j][index];
        }
        anom_pct[index] = min_score[index];
    }
    std::cout << "min_score" << std::endl;
    for (int i = 0; i < min_score.size(); i++) {
        if (min_score[i] < -500) {
            std::cout << "xxx " << min_score[i] << std::endl;
        }
        else {
                std::cout << min_score[i] << std::endl;
        }
    }
    std::cout << std::endl << std::endl;
    return std::make_tuple(top_indices, anom_pts, anom_scores, anom_pct, min_score);
}
